library/ssl_tls.c

```diff
@@ -7691,10 +7691,53 @@ unsigned int mbedtls_ssl_tls12_get_preferred_hash_for_sig_alg(
 
     for( i = 0; received_sig_algs[i] != MBEDTLS_TLS_SIG_NONE; i++ )
     {
-        if( sig_alg == MBEDTLS_SSL_TLS12_SIG_ALG_FROM_SIG_AND_HASH_ALG(
-                            received_sig_algs[i] ) )
-            return( MBEDTLS_SSL_TLS12_HASH_ALG_FROM_SIG_AND_HASH_ALG(
-                        received_sig_algs[i] ) );
+        unsigned int hash_alg_received =
+                    MBEDTLS_SSL_TLS12_HASH_ALG_FROM_SIG_AND_HASH_ALG(
+                        received_sig_algs[i] );
+        unsigned int sig_alg_received =
+                    MBEDTLS_SSL_TLS12_SIG_ALG_FROM_SIG_AND_HASH_ALG(
+                        received_sig_algs[i] );
+
+        if( sig_alg == sig_alg_received )
+        {
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+            if( ssl->handshake->key_cert && ssl->handshake->key_cert->key )
+            {
+                psa_algorithm_t alg = PSA_ALG_NONE;
+                psa_algorithm_t alg2 = PSA_ALG_NONE;
+                psa_key_usage_t usage = 0;
+                psa_key_usage_t usage2 = 0;
+
+                if( sig_alg_received == MBEDTLS_SSL_SIG_ECDSA )
+                {
+
+                    alg = PSA_ALG_ECDSA(
+                                mbedtls_psa_translate_md( hash_alg_received ) );
+                    usage = PSA_KEY_USAGE_SIGN_HASH;
+                    alg2 = PSA_ALG_ECDH;
```

Comment:
I don't think this alg2/usage2 thing is needed or correct: this function would only be called when the key will be used for signing according to the ciphersuite that was (tentatively) selected. Unless I'm mistaken, each ciphersuite only requires one use of the private key, so if we know it's signing, we know it's not the other (derive/decrypt).

Reply:
Indeed, I added it for both ECDSA & RSA for parity because the `MBEDTLS_KEY_EXCHANGE_RSA` ciphersuites failed if we only checked against `PSA_ALG_RSA_PKCS1V15_SIGN(hash_alg)`.
Also checking `PSA_ALG_RSA_PKCS1V15_CRYPT` fixes `MBEDTLS_KEY_EXCHANGE_RSA` ciphersuites, I added `PSA_ALG_ECDH` for parity but I was wrong.

Reply:
Ok, this is weird. I don't think we should be calling that function at all for `MBEDTLS_KEY_EXCHANGE_RSA` ciphersuites. This might be a pre-existing issue with the code. I'll double-check the existing code and the spec on Monday.

Reply:
This does the trick:
```c
diff --git a/library/ssl_ciphersuites.c b/library/ssl_ciphersuites.c
index 9b980234a..62988f259 100644
--- a/library/ssl_ciphersuites.c
+++ b/library/ssl_ciphersuites.c
@@ -1972,7 +1972,6 @@ mbedtls_pk_type_t mbedtls_ssl_get_ciphersuite_sig_alg( const mbedtls_ssl_ciphers
 {
     switch( info->key_exchange )
     {
-        case MBEDTLS_KEY_EXCHANGE_RSA:
         case MBEDTLS_KEY_EXCHANGE_DHE_RSA:
         case MBEDTLS_KEY_EXCHANGE_ECDHE_RSA:
             return( MBEDTLS_PK_RSA );
diff --git a/library/ssl_tls.c b/library/ssl_tls.c
index aa4aa0855..0abf2b99c 100644
--- a/library/ssl_tls.c
+++ b/library/ssl_tls.c
@@ -7713,13 +7713,10 @@ unsigned int mbedtls_ssl_tls12_get_preferred_hash_for_sig_alg(
                     continue;
 
                 if( sig_alg_received == MBEDTLS_SSL_SIG_RSA &&
-                    ! ( mbedtls_pk_can_do_ext( ssl->handshake->key_cert->key,
-                                               PSA_ALG_RSA_PKCS1V15_CRYPT,
-                                               PSA_KEY_USAGE_DECRYPT ) ||
-                        mbedtls_pk_can_do_ext( ssl->handshake->key_cert->key,
-                                               PSA_ALG_RSA_PKCS1V15_SIGN(
-                                                                psa_hash_alg ),
-                                               PSA_KEY_USAGE_SIGN_HASH ) ) )
+                    ! mbedtls_pk_can_do_ext( ssl->handshake->key_cert->key,
+                                             PSA_ALG_RSA_PKCS1V15_SIGN(
+                                                            psa_hash_alg ),
+                                             PSA_KEY_USAGE_SIGN_HASH ) )
                     continue;
             }
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
```

It seems `mbedtls_ssl_get_ciphersuite_sig_alg()` returns the hash alg if ciphersuite requires signing according to:
```
    /* If the ciphersuite requires signing, check whether
     * a suitable hash algorithm is present. */
```

Reply:
Indeed, I think `mbedtls_ssl_get_ciphersuite_sig_alg()` should return `MBEDTLS_PK_NONE` for `MBEDTLS_KEY_EXCHANGE_RSA` ciphersuites, as they don't use a signature. So, the patch above looks correct.